VoodooRMI/Transports/I2C/RMII2C.cpp

```diff
@@ -0,0 +1,335 @@
+
+    // maybe no need to strip the I2C (and RMI_READ_DATA_REPORT_ID) header?
+    u8 *i2cInput = new u8[len+4];
+    memset(databuff, 0, len);
+
+    IOLockLock(page_mutex);
+    if (RMI_I2C_PAGE(rmiaddr) != page) {
+        retval = rmi_set_page(RMI_I2C_PAGE(rmiaddr));
+        if (retval < 0)
+            goto exit;
+    }
+
+    if (device_nub->writeI2C(writeReport, sizeof(writeReport)) != kIOReturnSuccess) {
+        IOLog("%s: failed to read request output report\n", getName());
+        retval = -1;
+        goto exit;
+    }
+
+    if (device_nub->readI2C(i2cInput, len+4) != kIOReturnSuccess) {
+        IOLog("%s: failed to read I2C input\n", getName());
+        retval = -1;
+        goto exit;
+    }
+
+    if (i2cInput[2] != RMI_READ_DATA_REPORT_ID) {
+        IOLog("%s: RMI_READ_DATA_REPORT_ID mismatch\n", getName());
+        retval = -1;
+        goto exit;
+    }
+
+    memcpy(databuff, i2cInput+4, len);
+exit:
+//    IOLog("read %zd bytes at %#06x: %d (%*ph)\n", len, rmiaddr, ret, (int)len, databuff);
+    delete[] i2cInput;
+    IOLockUnlock(page_mutex);
+    return retval;
+}
+
+int RMII2C::blockWrite(u16 rmiaddr, u8 *buf, size_t len) {
+    int retval = 0;
+    u8 *writeReport = new u8[len+8] {
+        0x25,  // outputRegister & 0xFF; wOutputRegister
+        0x00,  // outputRegister >> 8;
+        (u8) ((len + 6) & 0xFF),  // size & 0xFF; 2 + reportID + buf (reportID excluded)
+        (u8) ((len + 6) >> 8),  // size >> 8;
+        RMI_WRITE_REPORT_ID,
+        (u8) len,
+        (u8) (rmiaddr & 0xFF),
+        (u8) (rmiaddr >> 8) };
+
+    IOLockLock(page_mutex);
+    if (RMI_I2C_PAGE(rmiaddr) != page) {
+        retval = rmi_set_page(RMI_I2C_PAGE(rmiaddr));
+        if (retval < 0)
+            goto exit;
+    }
+
+    memcpy(writeReport+8, buf, len);
+
+    if (device_nub->writeI2C(writeReport, len+8) != kIOReturnSuccess) {
+        IOLog("%s: failed to write request output report\n", getName());
+        retval = -1;
+        goto exit;
+    }
+    retval = 0;
+
+exit:
+//    IOLog("write %zd bytes at %#06x: %d (%*ph)\n", len, rmiaddr, ret, (int)len, buf);
+    delete [] writeReport;
+    IOLockUnlock(page_mutex);
+    return retval;
+}
+
+void RMII2C::interruptOccured(OSObject *owner, IOInterruptEventSource *src, int intCount) {
+    if (reading || !client)// || !awake)
+        return;
+
+    command_gate->attemptAction(OSMemberFunctionCast(IOCommandGate::Action, this, &RMII2C::notifyClient));
+}
+
+void RMII2C::notifyClient() {
+    // Do we really need it in command gate?
+    reading = true;
+    messageClient(kIOMessageVoodooI2CHostNotify, client);
+    reading = false;
+}
+
+void RMII2C::simulateInterrupt(OSObject* owner, IOTimerEventSource* timer) {
+    interruptOccured(owner, NULL, 0);
+    if (polling)
+        interrupt_simulator->setTimeoutMS(INTERRUPT_SIMULATOR_TIMEOUT);
+}
+
+bool RMII2C::setInterrupt(bool enable) {
+    // interrupt will only be enabled after client update
+    while (!client) {
```

Comment:
I'd be very hesistant to use a while loop. handleOpen should already have RMIBus (the client) in it, so I'd maybe use that (it sets a protected variable called `bus`). That said, I don't see it getting used in this function. Maybe could test to see if `bus` is set to Null or not. It'd be preferable if `bus` was used to reference RMIBus.